server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java

```diff
@@ -245,15 +245,28 @@ public synchronized void shutdownNode() {
 
         @Override
         public void run() {
+            int timeout = Integer.MAX_VALUE; // Default. Wait a very long time.
+            int expired = 0;
+            try {
+                timeout = ConfigParser.parse().getInt("nodeRecycleThreadWaitTimeout");
+            } catch (ConfigParserException e) { // NOSONAR
+                // ignore we have a default already
+            }
             appendMsgToCustomLog("Started NodeRecycleThread to restart the node once all the sessions are completed.");
-            while (getTotalUsed() > 0) {
+            while ((getTotalUsed() > 0) && (expired < timeout)) {
```

Comment:
I was about to point at this loop as suspect with any ill behaved tests.   Can we test for a possible hang here with a test?